File: backend/app/modules/reading_plan/routers/progress.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.modules.reading_plan.deps import get_current_user_id
from app.db import get_db
from app.modules.reading_plan.schemas.progress import (
    LibraryCommentListResponse,
    ProgressActivityListResponse,
    ProgressCreateRequest,
    ProgressCreateResponse,
    ProgressLogListResponse,
)
from app.modules.reading_plan.services.progress_service import (
    add_progress_log,
    list_library_comments,
    list_progress_activity,
    list_progress_logs,
)

router = APIRouter(prefix="/library", tags=["reading-plan:progress"])


@router.post("/{library_id}/progress", response_model=ProgressCreateResponse, status_code=201)
def add_progress_endpoint(
    library_id: int,
    payload: ProgressCreateRequest,
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user_id),
) -> ProgressCreateResponse:
    log, library = add_progress_log(
        db,
        library_id=library_id,
        user_id=user_id,
        page=payload.page,
        percent=payload.percent,
        memo=payload.memo,
    )
    return ProgressCreateResponse(log=log, library=library)


@router.get("/{library_id}/progress-logs", response_model=ProgressLogListResponse)
def list_progress_endpoint(
    library_id: int,
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user_id),
) -> ProgressLogListResponse:
    items = list_progress_logs(db, library_id=library_id, user_id=user_id)
    return ProgressLogListResponse(items=items)


@router.get("/comments", response_model=LibraryCommentListResponse)
def list_library_comments_endpoint(
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user_id),
) -> LibraryCommentListResponse:
    items = list_library_comments(db, user_id=user_id)
    return LibraryCommentListResponse(items=items)


@router.get("/progress-activity", response_model=ProgressActivityListResponse)
def list_progress_activity_endpoint(
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user_id),
) -> ProgressActivityListResponse:
    items = list_progress_activity(db, user_id=user_id)
    return ProgressActivityListResponse(items=items)

File: frontend/src/api/reading-plan/progress.ts
import type { LibraryComment, ProgressLogEntry, UserLibraryItem } from "../../types/reading-plan/book";
import { requestJson } from "./httpClient";

interface AddProgressInput {
  page?: number;
  percent?: number;
  memo?: string;
}

interface AddProgressResponse {
  log: ProgressLogEntry;
  library: UserLibraryItem;
}

interface ProgressLogListResponse {
  items: ProgressLogEntry[];
}

export async function addProgressLog(
  libraryId: string,
  input: AddProgressInput
): Promise<AddProgressResponse> {
  return requestJson<AddProgressResponse>(`/library/${libraryId}/progress`, {
    method: "POST",
    body: input,
  });
}

export async function fetchProgressLogs(libraryId: string): Promise<ProgressLogEntry[]> {
  const response = await requestJson<ProgressLogListResponse>(
    `/library/${libraryId}/progress-logs`
  );
  return response.items;
}

interface LibraryCommentListResponse {
  items: LibraryComment[];
}

export async function fetchLibraryComments(): Promise<LibraryComment[]> {
  const response = await requestJson<LibraryCommentListResponse>("/library/comments");
  return response.items;
}

interface ProgressActivityEntry {
  recordedAt: string;
}

interface ProgressActivityListResponse {
  items: ProgressActivityEntry[];
}

// 연속독서 계산용 — 진도 입력 시 코멘트를 남긴 날짜 목록(모든 서재 항목 통틀어서).
export async function fetchProgressActivity(): Promise<ProgressActivityEntry[]> {
  const response = await requestJson<ProgressActivityListResponse>("/library/progress-activity");
  return response.items;
}
